active/0051-strategies.md

```diff
@@ -0,0 +1,248 @@
+
+# Motivation
+
+The [Packager RFC](https://github.com/chadhietala/rfcs/blob/packager/active/0002-packager.md) submitted by [Chad Hietala](https://github.com/chadhietala) is a little over 2 years old. A lot of things have changed since then and it requires a revision.
+
+The current application build process merges and concatenates input broccoli trees. This behaviour is not well documented and is a tribal knowledge. While the simplicity of this approach is nice, it doesn't allow for extension. We can refactor our build process and provide more flexibility when desired.
+
+Most importantly, it helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember Engines)
++ unlock a whole different set of plugins we couldn't have before:
+  + ability to create custom bundles (i.e per-engine and per-route bundles)
+  + take advantage of [HTTP2 multiplexing](https://http2.github.io/faq/#why-is-http2-multiplexed) and [cache pushing](https://www.mnot.net/blog/2014/01/30/http2_expectations#4-cache-pushing)
+  + optimising plugins (Javascript and CSS treeshaking)
+
+
+
+# Scope
+
++ New public API for registering strategies
+
+# Terminology
+
++ **Strategy** - A strategy is responsible for returning a build pipeline that can emit a specific set of output assets given an input tree.
++ **Assembler** - Responsible for taking the resolved tree and applying – default or user provided – strategies to the tree.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a reader to understand.
+
+## Assembler
+
+`Assembler` is similar to a car assembly line - starting with a metal frame, making changes to it along the way, adding parts and at the end, there's an assembled car.
+
+It does not make any assumptions about how the final output should be constructed. Instead, it relies on strategies to tell it what the final output should look like. This is rather important as it enforces clear separation of concerns.
+
+Assembler must have the following interface:
+
+```typescript
+interface Assembler {
+  constructor(inputTree: BroccoliTree, strategies: Strategy[]);
+  toTree(): BroccoliTree;
+}
+```
+
+`toTree` method would gather all the trees from the passed in `strategies`, and merge them into one final broccoli tree.
+
+## Strategies
+
+Strategy is an extensibility primitive. It gives you granular control over the final output. It could be used in many different ways (we are going to go over use cases below). 
+
+Strategy must have the following interface:
+
+```typescript
+interface Strategy {
+  constructor(options: any);
+  toTree(assember: Assembler, inputTree: BroccoliTree): BroccoliTree;
```

Comment:
Curios: are there use cases where a `Strategy` needs access to the `Assembler`? Examples don't show this, and given the `Assembler` interface which only consists of constructor and `toTree`, looks like this could be omitted/further decoupled?

Reply:
Yes. Let's say we want to remove unused modules from an Ember application. `Assembler` could provide you with a list of modules/files that are used. That way it will be easy for a strategy to produce a broccoli tree with only relevant content. Some pseudocode (not a real implementation):

```javascript
const concat = require('broccoli-concat');

class DeadModulesEliminationStrategy() {
  constructor(options) {
    this.options = options;
  }

  toTree(assembler, inputTree) {
    return concat(inputTree, {
      // ... code ...
      inputFiles: assembler.getUsedModules()
      // ... code ...
    };
  }
}
```

Reply:
@twokul it seems odd to add a parameter solely as a bag of potential future API.  I had the same question as @simonihmig when reviewing.